### PR TITLE
Cassandra Driver Config Loader for Cassandra Driver Config in GCS

### PR DESCRIPTION
**Pull request by VardhanThigle**
# Overview
1. A common static utility to load [Cassandra Driver config ](https://docs.datastax.com/en/developer/java-driver/4.5/manual/core/configuration/index.html)from GCS.
2. The implementation uses the Cassandra Driver APIs to parse the file, so all the options that the driver supports (now and with upgrades) are supported for parsing out of the box.
3. The implementation also provides utility to serialize and deserialize the Config to [OptionMap](https://docs.datastax.com/en/drivers/java/4.9/com/datastax/oss/driver/api/core/config/OptionsMap.html)
## Testing
1. All lines covered in UT.
2. UT to parse the [reference config](https://docs.datastax.com/en/developer/java-driver/4.3/manual/core/configuration/reference/index.html) as provided by the Cassandra community.

### CODE REVIEW

File: v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/utils/CassandraDriverConfigLoader.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+      throw (new IllegalArgumentException(
+          String.format(
+              "Need to provide a single Cassandra driver config file in the specified path %s. Found %d files",
+              path, urls.length)));
+    }
+    return urls[0];
+  }
+
+  @VisibleForTesting
+  protected static void putOptionToMap(
```
Comment (bharadwaj-aditya) [inline v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/utils/CassandraDriverConfigLoader.java:123]: nit: Rename to putInOptionsMap or something. The current method rename sounds like the options map is created/added completely.

File: v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/utils/CassandraDriverConfigLoader.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+      getOptionsSupportedByDriver();
+
+  /**
+   * Load the Cassandra Config from a file as a {@link DriverConfigLoader}.
+   *
+   * @param path A complete gcs path to the config file of the form "gs://path/to/file".
+   * @return DriverConfigLoader.
+   * @throws FileNotFoundException - If file is not found at specified path.
+   */
+  public static DriverConfigLoader load(String path) throws FileNotFoundException {
```
Comment (bharadwaj-aditya) [inline v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/utils/CassandraDriverConfigLoader.java:52]: nit: suggest renaming to loadFile / loadFromPath or something which is more obvious, especially considering this is a public method.

File: v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/utils/CassandraDriverConfigLoader.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+   * java.io.Serializable}. At the time of opening a conneciton to Cassandra, it can be deserialized
+   * by {@link CassandraDriverConfigLoader#fromOptionsMap(OptionsMap)}. Note: Implementation Detail,
+   * Cassandra Driver does not provide a direct method to convert a link {@link DriverConfigLoader}
+   * into an {@link OptionsMap}, or build an {@link OptionsMap} from a file.
+   *
+   * @param path A complete gcs path to the config file of the form "gs://path/to/file".
+   * @return DriverConfigLoader.
+   * @throws FileNotFoundException - If file is not found at specified path.
+   */
+  public static OptionsMap loadToSerializableOptionsMap(String path) throws FileNotFoundException {
```
Comment (bharadwaj-aditya) [inline v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/utils/CassandraDriverConfigLoader.java:77]: nit: rename to be clearer - something like getOptionsMapFromFile or something.

File: v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/utils/CassandraDriverConfigLoader.java (hunk 2 above)
Comment (bharadwaj-aditya) [reply v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/utils/CassandraDriverConfigLoader.java:52]: does this method need to be public ? I think only 1 method will probably need to be public from this class - which is load/getOptionsMap

**Review by bharadwaj-aditya — COMMENTED** [review]
A few minor comments. Looks fine overall.

File: v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/utils/CassandraDriverConfigLoader.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+      LOG.error(
+          "Parsing error while parsing Cassandra Driver config from path {}", path, parseException);
+      throw parseException;
+    }
+  }
+
+  /**
+   * Load the Cassandra Config from a file as a {@link java.io.Serializable} {@link OptionsMap}.
+   * This {@link OptionsMap} can be stored in any object that needs to implement {@link
+   * java.io.Serializable}. At the time of opening a conneciton to Cassandra, it can be deserialized
```
[the remark this answers is not in the document]
Comment (VardhanThigle) [reply v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/utils/CassandraDriverConfigLoader.java]: Thanks.
Fixed in the current push.

File: v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/utils/CassandraDriverConfigLoader.java (hunk 1 above)
Comment (VardhanThigle) [reply v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/utils/CassandraDriverConfigLoader.java]: Thanks.
This method actually is `putOptionToOptionMap`, which sort of becomes redundant and hence had used `putOptionToMap`.
Changed this to `putInOptionsMap` as suggested. That works too.

File: v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/utils/CassandraDriverConfigLoader.java (hunk 2 above)
Comment (VardhanThigle) [reply v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/utils/CassandraDriverConfigLoader.java]: Thanks. Changed to `loadFile`.

The method is public to allow any pipeline to use loading to `DriverConfigLoader` directly if needed in their architecture at the time of pipeline construction.
I

File: v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/utils/CassandraDriverConfigLoader.java (hunk 3 above)
Comment (VardhanThigle) [reply v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/utils/CassandraDriverConfigLoader.java]: Thanks. Fixed in current push.

**Review by bharadwaj-aditya — APPROVED** [review]
LGTM

**Pull request merged**